mobile/android/android-virtual-device-service.ts

```diff
@@ -0,0 +1,283 @@
+import * as net from "net";
+import * as path from "path";
+import { EOL } from "os";
+import * as osenv from "osenv";
+import { AndroidVirtualDevice, DeviceTypes, NOT_RUNNING_EMULATOR_STATUS } from "../../constants";
+import { cache } from "../../decorators";
+
+export class AndroidVirtualDeviceService implements Mobile.IAndroidVirtualDeviceService {
+	private androidHome: string;
+
+	constructor(private $androidIniFileParser: Mobile.IAndroidIniFileParser,
+		private $childProcess: IChildProcess,
+		private $devicePlatformsConstants: Mobile.IDevicePlatformsConstants,
+		private $emulatorHelper: Mobile.IEmulatorHelper,
+		private $fs: IFileSystem,
+		private $hostInfo: IHostInfo) {
+			this.androidHome = process.env.ANDROID_HOME;
+		}
+
+	public async getAvailableEmulators(adbDevicesOutput: string[]): Promise<Mobile.IAvailableEmulatorsOutput> {
+		const availableEmulatorsOutput = await this.getAvailableEmulatorsCore();
+		const avds = availableEmulatorsOutput.devices;
+		const runningEmulatorIds = await this.getRunningEmulatorIds(adbDevicesOutput);
+		const runningEmulators = await Promise.all(runningEmulatorIds.map(emulatorId => this.getRunningEmulatorData(emulatorId, avds)));
+		const devices = availableEmulatorsOutput.devices.map(emulator => this.$emulatorHelper.getEmulatorByImageIdentifier(emulator.imageIdentifier, runningEmulators) || emulator);
+		return {
+			devices,
+			errors: availableEmulatorsOutput.errors
+		};
+	}
+
+	public async getRunningEmulatorIds(adbDevicesOutput: string[]): Promise<string[]> {
+		const emulatorIds = _.reduce(adbDevicesOutput, (result: string[], device: string) => {
+			const rx = device.match(AndroidVirtualDevice.RUNNING_AVD_EMULATOR_REGEX);
+			if (rx && rx[1]) {
+				result.push(rx[1]);
+			}
+
+			return result;
+		}, []);
+
+		return emulatorIds;
+	}
+
+	public startEmulator(imageIdentifier: string): void {
+		this.$childProcess.spawn(this.pathToEmulatorExecutable, ['-avd', imageIdentifier], { stdio: "ignore", detached: true }).unref();
```

Comment:
why is this unrefed and detached? And why do we just fire and forget about it? If it raises error, we'll not catch it as we use async spawn and it will be raised on the nextTick.